src/Aspire.Hosting.Azure.Functions/AzureFunctionsProjectResourceExtensions.cs

```diff
@@ -0,0 +1,150 @@
+    /// <param name="name"></param>
+    /// <returns></returns>
+    public static IResourceBuilder<AzureFunctionsProjectResource> AddAzureFunctionsProject<TProject>(this IDistributedApplicationBuilder builder, string name) where TProject : IProjectMetadata, new()
+    {
+        var projectDirectory = Path.GetDirectoryName(new TProject().ProjectPath)!;
+        var resource = new AzureFunctionsProjectResource(name, "func", projectDirectory);
+
+        // Add the default storage resource if it doesn't already exist.
+        var storage = builder.Resources.OfType<AzureStorageResource>().FirstOrDefault(r => r.Name == "azure-functions-default-storage");
+
+        if (storage is null)
+        {
+            storage = builder.AddAzureStorage("azure-functions-default-storage").RunAsEmulator().Resource;
+
+            builder.Eventing.Subscribe<BeforeStartEvent>((data, token) =>
+            {
+                var removeStorage = true;
+                // Look at all of the resources and if none of them use the default storage, then we can remove it.
+                // This is because we're unable to cleanly add a resource to the builder from within a callback.
+                foreach (var item in data.Model.Resources.OfType<AzureFunctionsProjectResource>())
+                {
+                    if (item.HostStorage == storage)
+                    {
+                        removeStorage = false;
+                    }
+
+                    // Before the resource starts, we want to apply the azure functions specific environment variables.
+                    // we look at all of the environment variables and apply the configuration for any resources that implement IResourceWithAzureFunctionsConfig.
+                    item.Annotations.Add(new EnvironmentCallbackAnnotation(static context =>
+                    {
+                        var functionsConfigMapping = new Dictionary<string, IResourceWithAzureFunctionsConfig>();
+                        var valuesToRemove = new List<string>();
+
+                        foreach (var (envName, val) in context.EnvironmentVariables)
+                        {
+                            var (name, config) = val switch
+                            {
+                                IResourceWithAzureFunctionsConfig c => (c.Name, c),
+                                ConnectionStringReference conn when conn.Resource is IResourceWithAzureFunctionsConfig c => (conn.ConnectionName ?? c.Name, c),
+                                _ => ("", null)
+                            };
+
+                            if (config is not null)
+                            {
+                                valuesToRemove.Add(envName);
+                                functionsConfigMapping[name] = config;
+                            }
+                        }
+
+                        // REVIEW: We need to remove the existing values before adding the new ones as there's a conflict with the connection strings.
```

Comment:
Is the conflict with the connection strings just the fact that the Funtions host expects connection strings in one format and the Aspire-injected resources don't but they both share the same resource names?

Reply:
Needs to be investigated. Something in the functions or extensions codebase attempts to look at the `ConnectionStrings__{connectionName}` and throws because it's looking for a specific template format (Key1=Value1;Key2=Value2)

Reply:
Tracked as a follow-up in https://github.com/dotnet/aspire/issues/5563.